woodwork/tests/accessor/test_accessor_utils.py

```diff
@@ -115,3 +118,84 @@ def test_get_valid_dtype(sample_series):
+            "The following columns in the typing information were missing from the DataFrame: {'id'}")
+
+    renamed_df = schema_df.rename(columns={'id': 'new_col'})
+    assert (get_invalid_schema_message(renamed_df, schema) ==
+            "The following columns in the DataFrame were missing from the typing information: {'new_col'}")
+
+
+def test_get_invalid_schema_message_dtype_mismatch(sample_df):
+    schema_df = sample_df.copy()
+    schema_df.ww.init(logical_types={'age': 'Categorical'})
+    schema = schema_df.ww.schema
+
+    incorrect_int_dtype_df = schema_df.ww.astype({'id': 'Int64'})
+    incorrect_bool_dtype_df = schema_df.ww.astype({'is_registered': 'Int64'})
+    incorrect_str_dtype_df = schema_df.ww.astype({'full_name': 'object'})  # wont work for koalas
+    incorrect_categorical_dtype_df = schema_df.ww.astype({'age': 'string'})  # wont work for koalas
+
+    assert (get_invalid_schema_message(incorrect_int_dtype_df, schema) ==
+            'dtype mismatch for column id between DataFrame dtype, Int64, and Integer dtype, int64')
+    assert (get_invalid_schema_message(incorrect_bool_dtype_df, schema) ==
+            'dtype mismatch for column is_registered between DataFrame dtype, Int64, and BooleanNullable dtype, boolean')
+    # Koalas backup dtypes make these checks not relevant
+    if ks and not isinstance(sample_df, ks.DataFrame):
+        assert (get_invalid_schema_message(incorrect_str_dtype_df, schema) ==
+                'dtype mismatch for column full_name between DataFrame dtype, object, and NaturalLanguage dtype, string')
+        assert (get_invalid_schema_message(incorrect_categorical_dtype_df, schema) ==
+                'dtype mismatch for column age between DataFrame dtype, string, and Categorical dtype, category')
+
+
+def test_get_invalid_schema_message_index_checks(sample_df):
+    if not isinstance(sample_df, pd.DataFrame):
+        pytest.xfail('Index validation not performed for Dask or Koalas DataFrames')
+
+    schema_df = sample_df.copy()
+    schema_df.ww.init(name='test_schema', index='id', logical_types={'id': 'Double', 'full_name': 'PersonFullName'})
+    schema = schema_df.ww.schema
+
+    different_underlying_index_df = schema_df.copy()
+    different_underlying_index_df['id'] = pd.Series([9, 8, 7, 6], dtype='float64')
+    assert (get_invalid_schema_message(different_underlying_index_df, schema) ==
+            "Index mismatch between DataFrame and typing information")
+
+    not_unique_df = schema_df.replace({3: 1})
+    not_unique_df.index = not_unique_df['id']
+    not_unique_df.index.name = None
+    assert get_invalid_schema_message(not_unique_df, schema) == 'Index column is not unique'
+
+
+def test_is_schema_valid_true(sample_df):
+    sample_df.ww.init(index='id', logical_types={'phone_number': 'Categorical'})
+    copy_df = sample_df.copy()
+
+    assert copy_df.ww.schema is None
+    assert is_schema_valid(copy_df, sample_df.ww.schema)
```

Comment:
Not sure I'm following you here - this is checking if the schema from sample_df (which is not `None` since WW is initialized on sample_df above) is valid for initializing on `copy_df` which has no schema and is just a copy of `sample_df` without the woodwork information.

Reply:
oh oops I definitely misread this! 

I guess, then, I have a separate question that arose from my pre-coffee blunder: What happens if you do `is_valid_schema(copy_df, None)`? Right now it raises an error, but I'm not sure that that's entirely accurate? And also I can imagine people checking `is_schema_valid(df, df.ww.schema)` if they need to confirm validity before performing some action but don't want to always have to confirm that there even is a schema initialized.

But it doesn't really make sense to me that you could get `is_schema_valid(df, None) is True`, so my suggestion if we want to handle the None case without an error was to switch up the direction we're checking here and have `is_schema_invalid` be what we check so we get `is_schema_invalid(df, None) is False`.

Reply:
(not that it's that big of an ask to have to do an `schema is not None` check before passing it to this function, but if we do leave it as is, we should probably add a test for this!)

Reply:
Hmmm. I guess was viewing this a little differently. We have a function that expects a DataFrame and a Schema as input arguments. Since `None` is not defined as valid for either input, users should see an error if they pass `None` for either one. I'm not sure it makes sense to handle the `None` input case here - in order to check if a schema is valid for a dataframe, users need to supply both of those things.

Now, having said that, it might be easy for users to supply `None` for the schema by mistake, and I guess it comes down to how many checks for valid input we want to put in place to try to give them a more helpful error message than the `AttributeError` they will currently get.

As you suggest, it's not a hard check to put in place, I'm just wrestling a bit with how much of this type of stuff we should try to catch for the user in general. I know we have done a fair amount in other places though.

Reply:
ok let's just leave it as is for now--we can always change later if the None case comes up as a big enough problem

Reply:
Sounds good!